File: src/python/bots/assistant_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os
import io
import json
import base64
import urllib.request

from botbuilder.core import ConversationState, TurnContext, UserState, MessageFactory
from botbuilder.schema import ChannelAccount, CardAction, ActionTypes, Activity
from botbuilder.dialogs import Dialog

from openai import AzureOpenAI
from openai.types.beta.assistant_stream_event import ThreadMessageDelta, ThreadRunRequiresAction, ThreadRunCreated, ThreadRunFailed
from openai.types.beta.threads import TextDeltaBlock, ImageFileDeltaBlock

from data_models import ConversationData, Attachment, mime_type
from .state_management_bot import StateManagementBot

class AssistantBot(StateManagementBot):

    def __init__(self, conversation_state: ConversationState, user_state: UserState, aoai_client: AzureOpenAI, dialog: Dialog):
        super().__init__(conversation_state, user_state, dialog)
        self.aoai_client = aoai_client
        self.chat_client = aoai_client.chat
        self.file_client = aoai_client.files

        self.deployment = os.getenv("AZURE_OPENAI_DEPLOYMENT_NAME")
        self.instructions = os.getenv("LLM_INSTRUCTIONS")
        self.welcome_message = os.getenv("LLM_WELCOME_MESSAGE", "Hello and welcome to the Assistant Bot Python!")
        self.assistant_id = os.getenv("AZURE_OPENAI_ASSISTANT_ID")
        self.streaming = os.getenv("AZURE_OPENAI_STREAMING", False)

    async def on_members_added_activity(self, members_added: list[ChannelAccount], turn_context: TurnContext):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity(self.welcome_message)

    async def on_message_activity(self, turn_context: TurnContext):
        
        # Enforce login
        loggedIn = await self.handle_login(turn_context)
        if not loggedIn:
            return False
        
        # Load conversation state
        conversation_data = await self.conversation_data_accessor.get(turn_context, ConversationData([]))

        # Create a new thread if one does not exist
        if conversation_data.thread_id is None:
            thread = self.aoai_client.beta.threads.create()
            conversation_data.thread_id = thread.id

        # Delete thread if user asks
        if turn_context.activity.text == 'clear':
            self.aoai_client.beta.threads.delete(conversation_data.thread_id)
            conversation_data.thread_id = None
            conversation_data.attachments = []
            conversation_data.history = []
            await turn_context.send_activity('Conversation cleared!')
            return True
                
        # Check if this is a file upload and process it
        files_uploaded = await self.handle_file_uploads(turn_context, conversation_data.thread_id, conversation_data)
        
        #  Return early if there is no text in the message
        if (turn_context.activity.text == None):
            return
    
        # Check if this is a file upload follow up - user selects a file upload option
        if (turn_context.activity.text.startswith(":")):
            # Get upload metadata
            tool = turn_context.activity.text.split(':').pop().strip()
            # Get file from attachments
            attachment = conversation_data.attachments[-1]
            # Add file upload to relevant tool
            with urllib.request.urlopen(attachment.url) as f:
                bytes = io.BytesIO(f.read())
                bytes.name = attachment.name
            file_response = self.file_client.create(file=bytes, purpose="assistants")
            # Send the file to the assistant
            tools = []
            if tool == "Code Interpreter":
                tools.append({
                    "type": "code_interpreter"
                })
            if tool == "File Search":
                tools.append({
                    "type": "file_search"
                })
            self.aoai_client.beta.threads.messages.create(
                thread_id=conversation_data.thread_id,
                role="user",
                content=f"File uploaded: {attachment.name}",
                attachments=[{
                    "file_id": file_response.id,
                    "tools": tools
                }]
            )
            # Send feedback to user
            await turn_context.send_activity(f"File added to {tool} successfully!");
            return True

        # Add user message to history
        conversation_data.add_turn("user", turn_context.activity.text)
        
        # Send user message to thread
        self.aoai_client.beta.threads.messages.create(
            thread_id=conversation_data.thread_id, 
            role="user", 
            content=turn_context.activity.text
        )
        
        # Run thread
        run = self.aoai_client.beta.threads.runs.create(
            thread_id=conversation_data.thread_id,
            assistant_id=self.assistant_id,
            instructions=self.instructions,
            stream=True
        )

        # Process run streaming
        await self.process_run_streaming(run, conversation_data, turn_context)

        return True

    async def process_run_streaming(self, run, conversation_data, turn_context, stream_id = None):
        # Start streaming response
        current_message = ""
        tool_outputs = []
        current_run_id = ""
        activity_id = ""
        stream_sequence = 1
        activity_id = await self.send_interim_message(turn_context, "Typing...", stream_sequence, stream_id, "typing")

        for event in run:
            if type(event) == ThreadRunFailed:
                current_message = event.data.last_error.message
                break
            if type(event) == ThreadRunCreated:
                current_run_id = event.data.id
            if type(event) == ThreadRunRequiresAction:
                tool_calls = event.data.required_action.submit_tool_outputs.tool_calls
                for tool_call in tool_calls:
                    arguments = json.loads(tool_call.function.arguments)
                    if tool_call.function.name == "image_query":
                        response = await self.image_query(conversation_data, arguments["query"], arguments["image_name"])
                        tool_outputs.append({"tool_call_id": tool_call.id, "output": response})

            if type(event) == ThreadMessageDelta:
                deltaBlock = event.data.delta.content[0]
                if type(deltaBlock) == TextDeltaBlock:
                    current_message += deltaBlock.text.value
                    stream_sequence += 1
                    # Flush content every 50 messages
                    if (stream_sequence % 50 == 0):
                        await self.send_interim_message(turn_context, current_message, stream_sequence, activity_id, "typing")

                elif type(deltaBlock) == ImageFileDeltaBlock: 
                    current_message += f"![{deltaBlock.image_file.file_id}](/api/files/{deltaBlock.image_file.file_id})"
        
        # Recursively process the run with the tool outputs
        if len(tool_outputs) > 0:
            new_run = self.aoai_client.beta.threads.runs.submit_tool_outputs(thread_id=conversation_data.thread_id, run_id=current_run_id, tool_outputs=tool_outputs, stream=True)
            await self.process_run_streaming(new_run, conversation_data, turn_context, activity_id)
            return
        response = current_message

        # Add assistant message to history
        conversation_data.add_turn("assistant", response)

        # Respond back to user
        stream_sequence += 1
        await self.send_interim_message(turn_context, current_message, stream_sequence, activity_id, "message")
    

    # Helper to handle file uploads from user
    async def handle_file_uploads(self, turn_context: TurnContext, thread_id: str, conversation_data: ConversationData):
        files_uploaded = False
        # Check if incoming message has attached files
        if turn_context.activity.attachments is not None:
            for attachment in turn_context.activity.attachments:
                if attachment.content_url == None:
                    continue
                files_uploaded = True
                download_url = attachment.content_url
                if attachment.content and "downloadUrl" in attachment.content:
                    download_url = attachment.content["downloadUrl"]
                # Add file to attachments in case we need to reference it in Function Calling
                conversation_data.attachments.append(Attachment(
                    name = attachment.name,
                    content_type = mime_type(attachment.name),
                    url = download_url
                ))

                # Add file upload notice to conversation history, frontend, and assistant
                conversation_data.add_turn("user", f"File uploaded: {attachment.name}")
                await turn_context.send_activity(f"File uploaded: {attachment.name}")
                self.aoai_client.beta.threads.messages.create(thread_id=thread_id,role="user",content=f"File uploaded: {attachment.name}",)
                # Ask whether to add file to a tool
                await turn_context.send_activity(MessageFactory.suggested_actions(
                    [
                        CardAction(title= ":Code Interpreter", type= ActionTypes.im_back, value= ":Code Interpreter"),
                        CardAction( title= ":File Search", type= ActionTypes.im_back, value= ":File Search"),
                    ],
                    "Add to a tool? (ignore if not needed)",
                ))


        # Return True if files were uploaded
        return files_uploaded
    
    async def image_query(self, conversation_data: ConversationData, query: str, image_name: str):
        # Find image in attachments by name
        image = next(filter(lambda a: a.name == image_name.split("/")[-1], conversation_data.attachments))
        # Handle image not found

        # Read image.url
        with urllib.request.urlopen(image.url) as f:
            # get file as base64
            bytes = base64.b64encode(f.read()).decode()

        # Send image to assistant
        response = self.chat_client.completions.create(
            model=self.deployment,
            messages=[
                {"role": "user", "content": [
                    {"type": "text", "text": query},
                    {"type": "image_url", "image_url": {
                        "url": f"data:{image.content_type};base64,{bytes}"}
                    }
                ]}
            ]
        )
        return response.choices[0].message.content

File: src/node/bots/assistant_bot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { MessageFactory, ActionTypes } = require('botbuilder');
const { ConversationData } = require('../data_models/conversation_data');
const { StateManagementBot } = require('./state_management_bot');
const { File } = require('node:buffer');
const { mimeType } = require('../data_models/mime_type');

class AssistantBot extends StateManagementBot {

    constructor(conversationState, userState, dialog, aoaiClient) {
        super(conversationState, userState, dialog);
        this.deployment = process.env.AZURE_OPENAI_DEPLOYMENT_NAME;
        this.aoaiClient = aoaiClient;
        this.welcomeMessage = process.env.LLM_WELCOME_MESSAGE || "Hello and welcome to the Chat Completions Bot NodeJS!";

        this.onMembersAdded(async (context, next) => {
            const membersAdded = context.activity.membersAdded;
            for (let member of membersAdded) {
                if (member.id !== context.activity.recipient.id) {
                    await context.sendActivity(this.welcomeMessage);
                }
            }
            await next();
        });

        this.onMessage(async (context, next) => {
            // Enforce login
            let loggedIn = await this.handleLogin(context)
            if (!loggedIn) {
                return false
            }
            if (context.activity.text === 'logout') {
                return await this.handleLogout(context)
            }

            // Load conversation state
            let conversationData = await this.conversationDataAccessor.get(context, new ConversationData([]));

            // Create a new thread if one does not exist
            let thread
            if (conversationData.thread_id === null) {
                thread = await this.aoaiClient.beta.threads.create();
                conversationData.thread_id = thread.id;
            } else {
                thread = await this.aoaiClient.beta.threads.retrieve(conversationData.thread_id);
            }

            // Delete thread if user asks
            if (context.activity.text === 'clear') {
                await this.aoaiClient.beta.threads.del(conversationData.thread_id);
                conversationData.thread_id = null;
                conversationData.attachments = [];
                conversationData.history = [];
                await context.sendActivity('Conversation cleared!');
                return true;
            }

            // Check if this is a file upload and process it
            let fileUploaded = await this.handleFileUploads(context, thread.id, conversationData);

            // Return early if there is no text in the message
            if (context.activity.text == null) {
                return true;
            }

            // Check if this is a file upload follow up - user selects a file upload option
            if (context.activity.text.startsWith(":")) {
                // Get upload metadata
                let tool = context.activity.text.split(':').pop().trim()
                // Get file from attachments
                let attachment = conversationData.attachments[conversationData.attachments.length - 1]
                // Send the file to the assistant
                let file = await fetch(attachment.url)
                let fileResponse = await this.aoaiClient.files.create({
                    file: new File([await file.blob()], attachment.name),
                    purpose: 'assistants'
                })
                // Add file upload to relevant tool
                let tools = []
                if (tool == "Code Interpreter")
                    tools.push({
                        "type": "code_interpreter"
                    })
                if (tool == "File Search")
                    tools.push({
                        "type": "file_search"
                    })
                await this.aoaiClient.beta.threads.messages.create(
                    conversationData.thread_id,
                    {
                        role: 'user',
                        content: `File uploaded: ${attachment.name}`,
                        attachments: [{
                            "file_id": fileResponse.id,
                            "tools": tools
                        }]
                    })
                // Send feedback to user
                await context.sendActivity(`File added to ${tool} successfully!`);
                return true
            }

            // Add user message to history
            ConversationData.addTurn(conversationData, 'user', context.activity.text);

            // Send user message to thread
            await this.aoaiClient.beta.threads.messages.create(
                conversationData.thread_id,
                {
                    role: 'user',
                    content: context.activity.text
                }
            );

            // Run thread
            let run = await this.aoaiClient.beta.threads.runs.create(
                conversationData.thread_id,
                {
                    assistant_id: process.env.AZURE_OPENAI_ASSISTANT_ID,
                    instructions: process.env.LLM_INSTRUCTIONS,
                    stream: true
                }
            );

            // Process run streaming
            await this.processRunStreaming(run, conversationData, context)

            next();

            return true
        });
    }


    async processRunStreaming(run, conversationData, context, streamId = null) {
        // Start streaming response
        let currentMessage = ""
        let toolOutputs = []
        let currentRunId = ""
        let activityId = ""
        let streamSequence = 1

        activityId = await this.sendInterimMessage(context, "Typing...", streamSequence, streamId, "typing")

        for await (let event of run.iterator()) {
            if (event.event == "thread.run.created")
                currentRunId = event.data.id
            if (event.event == "thread.run.failed") {
                currentMessage = event.data.last_error.message
                break
            }
            if (event.event == "thread.run.requires_action") {
                let tool_calls = event.data.required_action.submit_tool_outputs.tool_calls
                for (let tool_call of tool_calls) {
                    let args = JSON.parse(tool_call.function.arguments)
                    if (tool_call.function.name == "image_query") {
                        let response = await this.imageQuery(conversationData, args["query"], args["image_name"])
                        toolOutputs.push({ "tool_call_id": tool_call.id, "output": response })
                    }
                }

            }
            if (event.event == "thread.message.delta") {
                let deltaBlock = event.data.delta.content[0]
                if (deltaBlock.type == "text") {
                    currentMessage += event.data.delta.content[0].text.value;
                    // Flush content every 50 messages
                    if (streamSequence % 50 == 0) {
                        await this.sendInterimMessage(context, currentMessage, ++streamSequence, activityId, "typing")
                    }
                } else if (deltaBlock.type == "image_file") {
                    currentMessage += `![${deltaBlock.image_file.file_id}](/api/files/${deltaBlock.image_file.file_id})`
                }
            }
        }

        // Recursively process the run with the tool outputs
        if (toolOutputs.length > 0) {
            let newRun = await this.aoaiClient.beta.threads.runs.submitToolOutputs(
                conversationData.thread_id,
                currentRunId,
                { tool_outputs: toolOutputs, stream: true }
            )
            await this.processRunStreaming(newRun, conversationData, context, activityId)
            return
        }

        // Add assistant message to history
        ConversationData.addTurn(conversationData, 'assistant', currentMessage)

        // Respond back to user
        await this.sendInterimMessage(context, currentMessage, ++streamSequence, activityId, "message")
    }

    // Helper to handle file uploads from user
    async handleFileUploads(context, threadId, conversationData) {
        let filesUploaded = false;
        // Check if incoming message has attached files
        if (context.activity.attachments != null) {
            for (const attachment of context.activity.attachments) {
                if (attachment.contentUrl == null) {
                    continue;
                }
                filesUploaded = true;
                let downloadUrl = attachment.contentUrl;
                if (attachment.content && attachment.content.downloadUrl)
                    downloadUrl = attachment.content.downloadUrl
                // Add file to attachments in case we need to reference it in Function Calling
                conversationData.attachments.push({
                    name: attachment.name,
                    contentType: mimeType(attachment.name),
                    url: downloadUrl
                })

                // Add file upload notice to conversation history, frontend, and assistant
                ConversationData.addTurn(conversationData, 'user', `File uploaded: ${attachment.name}`)
                await context.sendActivity(`File uploaded: ${attachment.name}`)
                await this.aoaiClient.beta.threads.messages.create(
                    threadId,
                    {
                        role: 'user',
                        content: `File uploaded: ${attachment.name}`
                    }
                )
                // Ask whether to add file to a tool
                await context.sendActivity(MessageFactory.suggestedActions(
                    [
                        {
                            title: ":Code Interpreter",
                            type: ActionTypes.ImBack,
                            value: ":Code Interpreter"
                        },
                        {
                            title: ":File Search",
                            type: ActionTypes.ImBack,
                            value: ":File Search"
                        }
                    ],
                    "Add to a tool? (ignore if not needed)",
                ))
            }
        }

        // Return true if files were uploaded
        return filesUploaded
    }

    async imageQuery(conversationData, query, imageName) {
        // Find image in attachments by name
        let image = conversationData.attachments.find(a => a.name == imageName.split("/")[imageName.split("/").length - 1])

        // Read image.url
        let data = await fetch(image.url)
        let bytes = Buffer.from(await data.arrayBuffer()).toString('base64')

        // Send image to assistant
        let response = await this.aoaiClient.chat.completions.create({
            model: this.deployment,
            messages: [
                {
                    "role": "user", "content": [
                        { "type": "text", "text": query },
                        {
                            "type": "image_url", "image_url": {
                                "url": `data:${image.contentType};base64,${bytes}`
                            }
                        }
                    ]
                }
            ]
        })
        return response.choices[0].message.content

    }

}

module.exports.AssistantBot = AssistantBot;